src/lib.rs

```diff
@@ -418,242 +408,215 @@ impl Event {
     /// let event = Event::new();
-    /// let listener1 = event.listen();
-    /// let listener2 = event.listen();
+    /// let listener = event.listen();
     ///
-    /// assert!(listener1.same_event(&listener2));
+    /// // There are no notification so this times out.
+    /// assert!(!listener.wait_deadline(Instant::now() + Duration::from_secs(1)));
     /// ```
-    pub fn same_event(&self, other: &EventListener) -> bool {
-        ptr::eq::<Inner>(&*self.inner, &*other.inner)
+    #[cfg(not(loom))]
+    pub fn wait_deadline(self, deadline: Instant) -> bool {
+        self.wait_internal(Some(deadline))
     }
 
     fn wait_internal(mut self, deadline: Option<Instant>) -> bool {
-        // Take out the entry pointer and set it to `None`.
-        let entry = match self.entry.take() {
-            None => unreachable!("cannot wait twice on an `EventListener`"),
-            Some(entry) => entry,
-        };
-        let (parker, unparker) = parking::pair();
-
-        // Set this listener's state to `Waiting`.
-        {
-            let mut list = self.inner.lock();
-            let e = unsafe { entry.as_ref() };
-
-            // Do a dummy replace operation in order to take out the state.
-            match e.state.replace(State::Notified(false)) {
-                State::Notified(_) => {
-                    // If this listener has been notified, remove it from the list and return.
-                    list.remove(entry, self.inner.cache_ptr());
-                    return true;
-                }
-                // Otherwise, set the state to `Waiting`.
-                _ => e.state.set(State::Waiting(unparker)),
+        // Take out the entry.
+        let entry = self.entry.expect("already waited");
+
+        // Create a parker/unparker pair.
+        let (parker, unparker) = pair();
+
+        // Begin looping on `entry.wait()`.
+        let notified = loop {
+            let unparker = unparker.clone();
+
+            if Listener::wait(entry, move || Task::Thread(unparker)) {
+                // We have been notified.
+                break true;
             }
-        }
 
-        // Wait until a notification is received or the timeout is reached.
-        loop {
+            // Park the thread and see if we have been notified.
             match deadline {
-                None => parker.park(),
-
                 Some(deadline) => {
-                    // Check for timeout.
-                    let now = Instant::now();
-                    if now >= deadline {
-                        // Remove the entry and check if notified.
-                        return self
-                            .inner
-                            .lock()
-                            .remove(entry, self.inner.cache_ptr())
-                            .is_notified();
+                    #[cfg(loom)]
+                    {
+                        let _ = deadline;
+                        panic!("`wait_deadline` is not supported with loom");
```

Comment:
why do we not make this a compile-time error?

Reply:
This function is still used by the `wait()` function, even though the timeout in that case will always be `None`. A compiler error would prevent that from happening.

Reply:
Can we refactor it in such a way that `timeout` optionality is implemented via traits or such, instead of an `Option`, oh, and how does that interact with `no_std`, or do `wait` functions just not work there?

Reply:
I know this would "blow up" the API even more, but would imo be preferable.

Reply:
I could implement `timeout` as a trait object that selects between a `NoTimeout` option and a `Timeout(Duration)` option, but that feels like something that's out of scope for this PR.

Yes, there is no way to port the `wait()` functions to `no_std`, unless we wanted to try to port `parking` to `no_std`. IIRC that would be impossible unless we use [`critical-section`](https://crates.io/crates/critical-section) or something like that, and in my opinion, we shouldn't.